daliuge-engine/dlg/drop.py

```diff
@@ -164,6 +164,32 @@ class AbstractDROP(EventFirer, EventHandler):
     #  - Subclasses implement methods decorated with @abstractmethod
     __metaclass__ = ABCMeta
 
+    # Matcher used to validate environment_variable_syntax
+    _env_var_matcher = re.compile(r"\$[A-z|\d]+\..+")
+    _dlg_var_matcher = re.compile(r"\$DLG_.+")
+
+    _known_locks = ('_finishedProducersLock', '_refLock')
+    _known_rlocks = ('_statusLock',)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        for attr_name in AbstractDROP._known_locks + AbstractDROP._known_rlocks:
+            del state[attr_name]
+        del state["_listeners"]
+        return state
+
+    def __setstate__(self, state):
+        for attr_name in AbstractDROP._known_locks:
+            state[attr_name] = threading.Lock()
+        for attr_name in AbstractDROP._known_rlocks:
+            state[attr_name] = threading.RLock()
+
+        self.__dict__.update(state)
+
+        import collections
```

Comment:
can move to be a top-level import I think

Reply:
Moved :heavy_check_mark: